WifiART/src/userinterface/LciController.java

```diff
@@ -30,18 +41,66 @@ public LciController(LciView view, LciModel model) {
         this.view = view;
         this.model.setCallback(this);
 
-        view.addLatitudeListener(actionEvent ->
-            model.getState().setLatitude(view.getLatitude()));
-        view.addLatitudeUncertaintyListener(actionEvent ->
-            model.getState().setLatitudeUncertainty(view.getLatitudeUncertainty()));
-        view.addLongitudeListener(actionEvent ->
-            model.getState().setLongitude(view.getLongitude()));
-        view.addLongitudeUncertaintyListener(actionEvent ->
-            model.getState().setLongitudeUncertainty(view.getLongitudeUncertainty()));
-        view.addAltitudeListener(actionEvent ->
-            model.getState().setAltitude(view.getAltitude()));
-        view.addAltitudeUncertaintyListener(actionEvent ->
-            model.getState().setAltitudeUncertainty(view.getAltitudeUncertainty()));
+        view.addLatitudeListener(actionEvent -> {
```

Comment:
Note: to save yourself having to write lots of individual listeners and maintain them when adding future field additions,  you could have a general listener for any changes on a tab. The idea is that this triggers all fields to be read and the state updated using the current state of all fields on the tab. It occurs in the UI workflow, so the extra time spent each time you update a field is insignificant, but it makes the code more compact. Here is an example that I used for the BFA.  

  class UpdateDataSourceListener implements ActionListener {
    @Override
    public void actionPerformed(ActionEvent e) {
      SystemState state = fm.getState();
      try {
        state = fv.getViewState(state);
        if (e.getActionCommand() != null) {
          if (e.getActionCommand().equals("updateChartCommand")) {
            fm.updateChartWindow();
          }
        }
      } catch (NumberFormatException ex) {
        fv.displayErrorMessage(ERR_MESSAGE_NUMBER);
      } catch (IOException ex) {
        fv.displayErrorMessage(ERR_MESSAGE_FILE);
      } catch (ImagingOpException ex) {
        fv.displayErrorMessage(ERR_MESSAGE_IMAGE);
      }
      // Set new state in the View, or revert to the old state when an error occurred.
      fv.setViewState(state);
...

Reply:
This method makes sense, particularly for updating fields like checkboxes and radio buttons, rather than having individual listeners for each checkbox. That said, for text fields, which require input validation, I prefer the ability to validate the value of a specific field when the user interacts with it rather than validating all fields any time a change within the tab occurs. This is not impossible using the method you provided, as you can get the source of the action to determine which field to validate, but I think this particular usage is better encapsulated by independently defining the behavior for the field's listener.